### ISSUE TITLE
Add withDefault() support to MorphTo relationships 

### ISSUE THREAD

**Issue by vpratfr** [laravel/framework#24725]
Hi,

I am the person who originially opened #24055.

This is related to the fix in and its tests in #24061 which are not correct (unless I am wrong).

I fail to understand how calling `->withDefault()` could determine how to create the related model instance.

According to me, this should throw an exception because (if I follow the example used in the docs) the `Comment` class has no idea **what** to create as default (a `Video` or a `Post`?)

So the only valid use case of `withDefault` for `MorphTo` is when you provide a dynamic callback.

For instance: 

```
FooModality extends Model

BarModality extends Model

Mission extends Model  
    has attributes : 'template', 'modality_type', 'modality_id'

    public function modality() : MorphTo {
        return $this->morphTo('modality')
                ->withDefault(function() {
                    // Eloquent needs to update our 'modality_type', 'modality_id' columns
                    if ($this->template==='foo') return new FooModality();
                    if ($this->template==='bar') return new BarModality();

                    throw new IllegalArgumentException("Mission type is not handled");
                });
    }
```

If we want to be able to use `withDefault()` then we need to introduce somehow a mecanism to decide which class of related model to instanciate.

For instance a function with a given naming convention could be provided in the model which has the `MorphTo` relationship:

```
Mission extends Model

    // Function name is newDefault<<RelationshipName>>Instance
    protected function newDefaultModalityInstance() {
                    if ($this->template==='foo') return new FooModality();
                    if ($this->template==='bar') return new BarModality();

                    throw new IllegalArgumentException("Mission type is not handled");
    }

```

Then we could perfectly have `withDefault()` and `withDefault([...])` as valid use cases because they are able to know which class to instanciate.

**Comment by vpratfr** [comment]
In addition to the previous message, it is also worth mentionning that the withDefault callback does not receive the appropriate parameter

```
Mission extends Model  
    has attributes : 'template', 'modality_type', 'modality_id'

    public function modality() : MorphTo {
        return $this->morphTo('modality')
                ->withDefault(function($model) {

/* 
 * $model is a new Mission instance with empty attributes
 * This is not consistent with other withDefault implementations where it would receive the instance
 * of the related model, not the relationship owner model.
 */

                });
    }
```

**Comment by umair321** [comment]
not working in Laravel 5.6

**Comment by bucaneer** [comment]
Since #27411 was merged, I believe this issue is resolved. You can specify the morph type attribute (`modality_type` in your example) to force a particular class to be used for the related instance. Any conditional logic you need to choose the appropriate class can be put in a dynamic attribute getter:
```php
public function getModalityTypeAttribute($value = null) {
    if ($value === null) {
        if ($this->needsFoo()) return FooModality::class;
        if ($this->needsBar()) return BarModality::class;
    }
    return $value;
}
```

The callback passed to `withDefault` now receives the related model instance, so you can use it to customize the default attributes of the related model based on class, if you need to:
```php
public function modality() {
    return $this->morphTo()
        ->withDefault(function($model) {
            if ($model instanceof FooModality) $model->foo_attribute = 'something';
            $model->general_attribute = 'something else';
        });
}
```

**Comment by taylorotwell** [comment]
Your PR will be reverted. Causes failing tests on master.

**Comment by driesvints** [comment]
I'll re-open this so I can have a look later at some point.

**Comment by vpratfr** [comment]
Thanks.

I am struggling with how we are supposed to used the current code **if** that is meant to be sufficient to cover the needs for default morph related models.

Closest would be to use the received parent model in order to return a new instance of the morphed related model.

```
public function modality() {
    return $this->morphTo()->withDefault(function($parent) {
        if ($parent->needsFoo()) return new Foo();
        if ($parent->needsBar()) return new Bar();
        return null;
    });
}
```

In that case:

- are we supposed to associate the new Foo or Bar models with the parent?
- is that done by Eloquent automatically?

Seems a very different behaviour than `withDefault` when used on non-polymorphic relationships: those callbacks directly receive an instanciated and linked related model instance.

**Comment by taylorotwell** [comment]
This was fixed by #27455
